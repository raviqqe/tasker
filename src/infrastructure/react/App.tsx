import { PulseLoader } from "react-spinners";
import { useAsync } from "react-use";
import React from "react";
import styled from "styled-components";
import { observer } from "mobx-react";
import { AuthenticationStore } from "../mobx/authentication-store";
import { DocumentsStore } from "../mobx/documents-store";
import { IProps as ILandingProps, Landing } from "./Landing";
import { Home, IProps as IHomeProps } from "./Home";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
`;

interface IProps extends Omit<IHomeProps, "documents">, ILandingProps {
  authenticationStore: AuthenticationStore;
  documentsStore: DocumentsStore;
  initialize: () => Promise<void>;
}

export const App = observer(
  ({
    authenticationStore: { signedIn },
    documentsStore: { documents },
    initialize,
    repositoryURL,
    signIn,
    signOut,
    ...props
  }: IProps) => {
    useAsync(initialize, []);

    return signedIn === null ? (
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
    ) : signedIn ? (
      <Home {...props} documents={documents} signOut={signOut} />
    ) : (
      <Landing repositoryURL={repositoryURL} signIn={signIn} />
    );
  }
);
