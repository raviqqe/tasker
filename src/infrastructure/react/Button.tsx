import styled from "styled-components";
import { boxShadow } from "./style";
import { red, grey, white } from "./style/colors";

export const Button = styled.button<{ secondary?: boolean }>`
  ${boxShadow};
  background: ${({ secondary }) => (secondary ? grey : red)};
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: inherit;
  color: ${white};
  border: none;
  cursor: pointer;
  flex-shrink: 0;
`;
