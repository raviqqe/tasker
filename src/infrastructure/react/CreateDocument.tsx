import { MdAdd } from "react-icons/md";
import React, { useState } from "react";
import styled from "styled-components";
import { CircleButton } from "./CircleButton";
import { InsertFilesFunction } from "./utilities";
import { MarkdownTextArea } from "./MarkdownTextArea";

const Container = styled.div`
  display: flex;
  align-items: center;
  padding: 1em;
`;

const StyledMarkdownTextArea = styled(MarkdownTextArea)`
  margin-right: -1em;
  max-height: 80vh;
`;

interface IProps {
  createDocument: (text: string) => Promise<void>;
  insertFiles: InsertFilesFunction;
}

export const CreateDocument = ({ createDocument, insertFiles }: IProps) => {
  const [text, setText] = useState("");
  const onSubmit = async (): Promise<void> => {
    setText("");
    await createDocument(text);
  };

  return (
    <Container>
      <StyledMarkdownTextArea
        insertFiles={insertFiles}
        onSubmit={onSubmit}
        setText={setText}
        text={text}
      />
      <CircleButton aria-label="Create" onClick={onSubmit}>
        <MdAdd />
      </CircleButton>
    </Container>
  );
};
