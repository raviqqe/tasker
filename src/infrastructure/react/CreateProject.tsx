import { MdAdd } from "react-icons/md";
import React from "react";
import { CircleButton } from "./CircleButton";

export interface IProps {
  createProject: (name: string) => Promise<void>;
}

export const CreateProject = ({ createProject, ...restProps }: IProps) => (
  <CircleButton
    aria-label="Create Project"
    onClick={async () => {
      const name = window.prompt("Project name?");

      if (!name) {
        return;
      }

      await createProject(name);
    }}
    {...restProps}
  >
    <MdAdd />
  </CircleButton>
);
