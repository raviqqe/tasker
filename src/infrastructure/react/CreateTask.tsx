import { MdAdd } from "react-icons/md";
import React from "react";
import { CircleButton } from "./CircleButton";

interface IProps {
  createTask: (name: string) => Promise<void>;
}

export const CreateTask = ({ createTask }: IProps) => {
  return (
    <CircleButton
      aria-label="Create"
      onClick={async () => {
        const name = window.prompt("Task name?");

        if (!name) {
          return;
        }

        await createTask(name);
      }}
    >
      <MdAdd />
    </CircleButton>
  );
};
