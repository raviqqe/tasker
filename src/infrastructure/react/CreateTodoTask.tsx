import { MdAdd } from "react-icons/md/index.js";
import { CircleButton } from "./CircleButton.js";

export interface Props {
  className?: string;
  createTodoTask: (name: string) => Promise<void>;
}

export const CreateTodoTask = ({
  createTodoTask,
  ...restProps
}: Props): JSX.Element => (
  <CircleButton
    aria-label="Create"
    onClick={async () => {
      const name = window.prompt("Task name?");

      if (!name) {
        return;
      }

      await createTodoTask(name);
    }}
    {...restProps}
  >
    <MdAdd />
  </CircleButton>
);
