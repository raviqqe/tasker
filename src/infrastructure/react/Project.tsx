import { styled } from "@linaria/react";
import { forwardRef, type Ref } from "react";
import {
  MdArchive,
  MdUnarchive,
  MdDelete,
  MdEdit,
} from "react-icons/md/index.js";
import type * as domain from "../../domain.js";
import { IconButton } from "./IconButton.js";
import { black, red } from "./style/colors.js";

const Container = styled.div`
  display: flex;
  justify-content: space-between
  align-items: center;
  font-size: 1.25em;
`;

const Name = styled.div<{ highlighted: boolean }>`
  word-break: break-word;
  margin-right: 1em;
  flex: 1;
  cursor: ${({ onClick }) => (onClick ? "pointer" : "auto")};
  color: ${({ highlighted }) => (highlighted ? red : black)};
`;

const ButtonsContainer = styled.div`
  display: flex;
  align-items: center;

  > :not(:first-child) {
    margin-left: 0.25rem;
  }
`;

export interface Props {
  archiveProject?: (
    project: domain.Project,
    currentProjectId: string,
  ) => Promise<void>;
  currentProject?: domain.Project;
  deleteProject?: (project: domain.Project) => Promise<void>;
  project: domain.Project;
  switchCurrentProject?: (project: domain.Project) => Promise<void>;
  unarchiveProject?: (project: domain.Project) => Promise<void>;
  updateProject?: (project: domain.Project) => Promise<void>;
}

const ProjectWithRef = (
  {
    archiveProject,
    currentProject,
    deleteProject,
    project,
    switchCurrentProject,
    unarchiveProject,
    updateProject,
  }: Props,
  ref: Ref<HTMLDivElement>,
) => (
  <Container ref={ref}>
    <Name
      highlighted={!!currentProject && project.id === currentProject.id}
      onClick={switchCurrentProject && (() => switchCurrentProject(project))}
    >
      {project.name}
    </Name>
    <ButtonsContainer>
      {updateProject && (
        <IconButton
          aria-label="Update Project"
          onClick={async () => {
            const name = window.prompt("New project name?", project.name);

            if (!name) {
              return;
            }

            await updateProject({ ...project, name });
          }}
        >
          <MdEdit />
        </IconButton>
      )}
      {archiveProject && (
        <IconButton
          aria-label="Archive Project"
          onClick={async () => {
            if (currentProject) {
              await archiveProject(project, currentProject.id);
            }
          }}
        >
          <MdArchive />
        </IconButton>
      )}
      {unarchiveProject && (
        <IconButton
          aria-label="Unarchive Project"
          onClick={() => unarchiveProject(project)}
        >
          <MdUnarchive />
        </IconButton>
      )}
      {deleteProject && (
        <IconButton
          aria-label="Delete Project"
          onClick={() => deleteProject(project)}
        >
          <MdDelete />
        </IconButton>
      )}
    </ButtonsContainer>
  </Container>
);

export const Project = forwardRef(ProjectWithRef);
