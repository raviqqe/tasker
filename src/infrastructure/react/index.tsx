import { render } from "react-dom";
import React from "react";
import { ApplicationInitializer } from "../../application/application-initializer";
import { DocumentCreator } from "../../application/document-creator";
import { DocumentLister } from "../../application/document-lister";
import { DocumentUpdater } from "../../application/document-updater";
import { IDocument } from "../../domain/document";
import { SignInManager } from "../../application/sign-in-manager";
import { SignOutManager } from "../../application/sign-out-manager";
import { TextFileInserter } from "../../application/text-file-inserter";
import { AuthenticationStore } from "../mobx/authentication-store";
import { DocumentsStore } from "../mobx/documents-store";
import { GlobalStyle } from "./style";
import { App } from "./App";

export class ReactRenderer {
  constructor(
    private readonly applicationInitializer: ApplicationInitializer,
    private readonly documentCreator: DocumentCreator,
    private readonly documentLister: DocumentLister,
    private readonly documentUpdater: DocumentUpdater,
    private readonly signInManager: SignInManager,
    private readonly signOutManager: SignOutManager,
    private readonly textFileInserter: TextFileInserter,
    private readonly authenticationStore: AuthenticationStore,
    private readonly documentsStore: DocumentsStore,
    private readonly repositoryURL: string
  ) {}

  public render(element: HTMLElement): void {
    render(
      <>
        <App
          authenticationStore={this.authenticationStore}
          documentsStore={this.documentsStore}
          createDocument={(text: string) => this.documentCreator.create(text)}
          initialize={() => this.applicationInitializer.initialize()}
          insertFiles={(
            text: string,
            position: number,
            files: File[]
          ): Promise<string> =>
            this.textFileInserter.insert(text, position, files)
          }
          listDocuments={() => this.documentLister.list()}
          listMoreDocuments={() => this.documentLister.listMore()}
          repositoryURL={this.repositoryURL}
          signIn={() => this.signInManager.signIn()}
          signOut={() => this.signOutManager.signOut()}
          updateDocument={(document: IDocument, text: string) =>
            this.documentUpdater.update(document, text)
          }
        />
        <GlobalStyle />
      </>,
      element
    );
  }
}
