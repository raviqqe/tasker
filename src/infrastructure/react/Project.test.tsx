import { render } from "@testing-library/react";
import { it, expect } from "vitest";
import { Project } from "./Project.js";

it("renders", () => {
  expect(
    render(<Project project={{ archived: false, id: "", name: "" }} />)
      .container.firstChild,
  ).toMatchSnapshot();
});
