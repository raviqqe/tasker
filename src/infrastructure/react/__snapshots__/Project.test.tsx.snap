// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1`] = `
<div
  class="mocked-styled-1 c1g0fyep"
>
  <div
    class="mocked-styled-2 n1wknujg"
    style="--n1wknujg-0: auto; --n1wknujg-1: #222;"
  />
  <div
    class="mocked-styled-3 b12yr94b"
  />
</div>
`;
